### DISCUSSION TITLE
IN operator

### DISCUSSION THREAD

**Discussion by curquiza** [meilisearch/product#407] [Feedback & Feature Proposal]
Related issue: https://github.com/meilisearch/meilisearch/issues/2163

The users seem to need the IN operator when searching with `filter`.

If you are a Meilisearch user, feel free to give more detail about use case and your need :)

**Comment by francoism90** [comment]
This is really useful when you need to filter on user feeds. For example, a user that search on his/her bookmarks.

**Comment by eelco-weblog** [comment]
This feature can also be useful for a related feature:
https://github.com/meilisearch/product/discussions/77

For example an e-commerce site wants to display Apple Airpods, but not Apple Airpods Max. Now, we can create an array of all the words in the title, save it as another property such as `title_words` and filter something like this:

`(title_words = airpod and title_words != max)`

When we can use CONTAINS / NOT CONTAINS, we do not have to create another property and filter directly on the title:

`(title CONTAINS airpod and title NOT CONTAINS max)`

When this is implemented, it might also be a solution for the use-cases for this discussion: https://github.com/meilisearch/product/discussions/77

**Reply by curquiza** [reply]
Hello @francoism90 and @eelco-weblog,
Thanks for your feedback! FYI, we are considering adding this feature :)

**Reply by francoism90** [reply]
@curquiza Great! 👍

**Comment by JerryBels** [comment]
I got here by searching for a solution for a NOT IN use case. I'm making a multiple users selection tool, and would like the search to filter out already selected users from the list. A `NOT IN` filter would be ideal for that. For the time being I will use a loop with `!=` filters.

**Comment by gmourier** [comment]
A technical oriented discussion now exists [here ](https://github.com/meilisearch/product/discussions/479)

**Comment by gmourier** [comment]
Hello dear community, 👋

We have published a docker image allowing you to test the addition of the `IN` operator.

### `IN` operator

Instead of writing a cumbersome clause like `user_id = 1 OR user_id = 2 OR user_id = 3 ...`. It's possible to write `user_id IN [1, 2, 3]` and voila!

- `user_id IN [...]` selects all documents whose attribute contains at least one of the values given inside the brackets. In case there are no values, then the filter selects no documents.
- `NOT user_id IN[...]` selects all documents whose attribute contains values that are all different from those in the brackets. (It includes documents missing the attribute - See `EXISTS` operator for more granularity).

---

We have also made some other changes that may be of interest to you.

### `NOT` / `!=` behavior

`NOT` /`!=` operator has been enhanced following this [discussion](https://github.com/meilisearch/product/discussions/325).

`tags != adult` now returns documents having `tags: null` or `tags: []` but also documents without the `tags` property. It's equivalent to `NOT tags = adult`.

The `NOT` operator selects all documents not selected by the filter that follows it.

### `EXISTS` operator

We took the opportunity to add the operator `EXISTS`. This operator allows you to granularly require that the selected documents have a `tag` attribute regardless of its value.

In this case, `tags != adult and tags EXISTS` to retrieve documents that do not contain the `adult` value in the `tag` attribute and only those for which the `tag` attribute exists.

- `tags EXISTS` selects all documents having the `tags` attribute.
- `NOT tags EXISTS` selects all documents missing the `tags` attribute.

---

We would like to have your feedback before deciding whether to introduce the given changes in a future version.

To do so you have to pull the docker image we prepared.

```bash
docker run -it --rm \
    -p 7700:7700 \
    -v $(pwd)/meili_data:/meili_data \
    getmeili/meilisearch:v0.29.0-filter.beta.0
```

Thank you so much for being part of our community ✨

**Reply by francoism90** [reply]
@gmourier I'm using Laravel Scout - any (beta) packages I should try?

Thanks for this! It will speed up querying a lot I think. :)

**Reply by brunoocasali** [reply]
Hi @francoism90 meilisearch-php is not compatible with Meilisearch v0.28 yet! But it will be in the next week - on Monday :) 
As soon as the Meilisearch engine releases its new version we will release the new version of our SDKs, so after that, you can try these new awesome filters in your laravel app 💪

**Reply by gmourier** [reply]
I know it's not ideal @francoism90 but you can test it with curl or postman if you really want to do it before next week. 🏃‍♂️

**Comment by gmourier** [comment]
Hello 👋
We have just released [v0.29.0rc1](https://github.com/meilisearch/meilisearch/releases/tag/v0.29.0rc1), which is a release candidate of v0.29.0 🔥

You can test the just-introduced `EXISTS` and `IN` filtering operators!
More information in [this issue](https://github.com/meilisearch/meilisearch/issues/2479) and [this one](https://github.com/meilisearch/meilisearch/issues/2484).

[Binaries](https://github.com/meilisearch/meilisearch/releases/tag/v0.29.0rc1) are attached to the release, or you can use the docker image:
```bash
docker run -it --rm \
    -p 7700:7700 \
    getmeili/meilisearch:v0.29.0rc1
```

Let us know about any bugs or feedback! 😄 It would be really helpful.

FYI, the official v0.29.0 release will be available on 3rd October.

**Comment by gmourier** [comment]
Hey everyone 👋

Locking this discussion since the `IN` operator has been released.

`IN [valueA, valueB, …, valueN]` selects all documents whose chosen field contains at least one of the specified values. The expression `genres IN [horror, comedy]` returns all documents whose genres includes either horror, comedy, or both

📚 https://docs.meilisearch.com/learn/advanced/filtering_and_faceted_search.html#using-filters

Please open a new discussion if you have any feedback to share 🙇‍♂️
